mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/MlflowAutologEventPublisher.scala

```diff
@@ -0,0 +1,130 @@
+  }
+
+  // Initialize Spark listener that pulls Delta query plan information & bubbles it up to registered
+  // Python subscribers, along with a GC loop for removing unrespoins
+  def init(gcDeadSubscribersIntervalSec: Int = 1): Unit = synchronized {
+    if (sparkQueryListener == null) {
+      val listener = getSparkDataSourceListener
+      // NB: We take care to set the variable only after adding the Spark listener succeeds,
+      // in case listener registration throws. This is defensive - adding a listener should
+      // always succeed.
+      spark.sparkContext.addSparkListener(listener)
+      sparkQueryListener = listener
+      // Schedule regular cleanup of detached subscribers, e.g. those associated with detached
+      // notebooks
+      val task = new Runnable {
+        def run(): Unit = {
+          unregisterBrokenSubscribers()
+        }
+      }
+      scheduledTask = ex.scheduleAtFixedRate(task, 1, gcDeadSubscribersIntervalSec, TimeUnit.SECONDS)
+    }
+  }
+
+  def stop(): Unit = synchronized {
+    if (sparkQueryListener != null) {
+      spark.sparkContext.removeSparkListener(sparkQueryListener)
+      sparkQueryListener = null
+      while(!scheduledTask.cancel(false)) {
+        Thread.sleep(1000)
+      }
+      subscribers = mutable.LinkedHashMap()
+    }
+  }
+
+  def register(subscriber: MlflowAutologEventSubscriber): String = synchronized {
+    if (sparkQueryListener == null) {
+      throw new RuntimeException("Please call init() before attempting to register a subscriber")
+    }
+    this.synchronized {
+      val uuid = java.util.UUID.randomUUID().toString
+      subscribers.put(uuid, subscriber)
+      uuid
+    }
+  }
+
+  /** Unregister subscribers broken e.g. due to detaching of the associated Python REPL */
+  private[autologging] def unregisterBrokenSubscribers(): Unit = synchronized {
+    val brokenUuids = subscribers.flatMap { case (uuid, listener) =>
+      try {
+        listener.ping()
+        Seq.empty
+      } catch {
+        case e: Py4JException =>
+          logger.info(s"Subscriber with UUID $uuid not responding to health checks, removing it")
+          Seq(uuid)
+        case NonFatal(e) =>
+          logger.error(s"Unexpected exception while checking health of subscriber $uuid, " +
+            s"removing it. Please report this error at https://github.com/mlflow/mlflow/issues, " +
+            s"along with the following stacktrace:\n" +
+            s"${ExceptionUtils.serializeException(e)}")
+          Seq(uuid)
+      }
+    }
+    brokenUuids.foreach { uuid =>
+      subscribers.remove(uuid)
+    }
+  }
+
+  private[autologging] def publishEvent(
+      replIdOpt: Option[String],
+      sparkTableInfo: SparkTableInfo): Unit = synchronized {
+    sparkTableInfo match {
+      case SparkTableInfo(path, version, format) =>
+        for ((uuid, listener) <- subscribers) {
+          try {
+            if (replIdOpt.isEmpty || listener.replId == replIdOpt.get) {
```

Comment:
when is replIdOpt empty?

Reply:
Good question it's actually currently always empty, as we don't know which REPL triggered the datasource-read event that is being published in `publishEvent`, and so we publish to all subscribers.

In REPL-ID aware environments (e.g. Databricks), we will be able to pull the REPL ID associated with the query in our SparkListener, so this logic was added in anticipation of that use case - will remove it for now & reintroduce it with a follow-up PR that adds a REPL-ID aware SparkListener